### PR TITLE
Get secret/configmaps information from config for all IdPs

### PR DESCRIPTION
**Pull request by stlaz**
Grabs identity providers' configmaps and secrets from the
top-level config and stores their mapping to volumes/volumeMounts
so that these can later be added to the deployment running OSIN

### CODE REVIEW

**Comment by stlaz** [comment]
Note to self: get rid of `idpSyncData`, it does not seem to be bringing any value, only overhead

File: pkg/operator2/configsync.go [hunk 1]
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
...
+// TODO: new source data could be generalized
+func newSourceDataIDPSecret(index int, secretName, idpType string) (string, sourceData) {
+	dest := getIDPName(index, secretName, idpType)
+	var ret sourceData
+
+	volume, mount, path := secretVolume(index, dest, idpType)
+	ret.src = secretName
+	ret.path = path
+	ret.volume = volume
+	ret.mount = mount
```
Comment (ericavonb) [inline pkg/operator2/configsync.go]: Nit: just do 
```go
ret := sourceData{
	src: secretName,
	path: path,
	volume: volume,
	mount: mount,
}
```

File: pkg/operator2/configsync.go [hunk 2]
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
...
+
+func newSourceDataIDPConfigMap(index int, cmName, idpType string) (string, sourceData) {
+	dest := getIDPName(index, cmName, idpType)
+	var ret sourceData
+
+	volume, mount, path := secretVolume(index, dest, idpType)
+	ret.src = cmName
+	ret.path = path
+	ret.volume = volume
+	ret.mount = mount
```
Comment (ericavonb) [inline pkg/operator2/configsync.go]: Same nit about removing var and mutations in favor of setting in new struct directly

File: pkg/operator2/configsync.go [hunk 3]
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
...
+
+			out = append(out,
+				idpSyncData{
+					configMaps: map[string]sourceData{clientCADest: clientCAData},
+				},
+			)
+
 		default:
-			panic("TODO")
+			return nil // TODO: some erroring
```
Comment (ericavonb) [inline pkg/operator2/configsync.go]: `return fmt.Errorf(“unknown IdentityProvider type: %s”, pc.Type)`

File: pkg/operator2/configsync.go [hunk 4]
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
...
+			p := pc.BasicAuth
+
+			ca := p.CA.Name
+			caDest, caData := newSourceDataIDPConfigMap(i, ca, corev1.ServiceAccountRootCAKey)
+
+			clientCert := p.TLSClientCert.Name
+			clientCertDest, clientCertData := newSourceDataIDPSecret(i, clientCert, corev1.TLSCertKey)
+
+			clientKey := p.TLSClientKey.Name
+			clientKeyDest, clienKeyData := newSourceDataIDPSecret(i, clientKey, corev1.TLSCertKey)
```
Comment (ericavonb) [inline pkg/operator2/configsync.go]: Should be `corev1.TLSPrivateKeyKey`

File: pkg/operator2/configsync.go [hunk 5]
```diff
@@ -91,64 +91,189 @@ type sourceData struct {
...
 // TODO this should be combined with convertProviderConfigToOsinBytes as it would simplify how the data is shared
 func convertToData(idps []configv1.IdentityProvider) []idpSyncData {
 	out := make([]idpSyncData, 0, len(idps))
 	for i, idp := range idps {
 		pc := idp.ProviderConfig
 		switch pc.Type {
+		case configv1.IdentityProviderTypeBasicAuth:
+			p := pc.BasicAuth
+
+			ca := p.CA.Name
```
Comment (ericavonb) [inline pkg/operator2/configsync.go]: nit: unnecessary var declaration. Use `p.CA.Name` directly. Same for `clientCert` and `clientKey`

File: pkg/operator2/configsync.go (hunk 3 above)
Comment (ericavonb) [reply pkg/operator2/configsync.go]: And thank you, no panics!

File: pkg/operator2/configsync.go (hunk 4 above)
Comment (stlaz) [reply pkg/operator2/configsync.go]: Good catch, I was afraid something might have gotten lost in those copy-pastes.

File: pkg/operator2/configsync.go (hunk 3 above)
Comment (stlaz) [reply pkg/operator2/configsync.go]: The function currently only returns `[]idpSyncData`, I'd maybe fix this when gets merged with `convertProviderConfigToOsinBytes()`

**Comment by stlaz** [comment]
Fixed comments, rebased on master.

I also added a refactoring to `convertToData()` which now only returns a pointer to the `idpSyncData` structure. Lets' hope nothing breaks!

File: pkg/operator2/configsync.go (hunk 3 above)
Comment (ericavonb) [reply pkg/operator2/configsync.go]: icky, ok. for another pr

File: pkg/operator2/configsync.go [hunk 6]
```diff
@@ -91,67 +90,133 @@ type sourceData struct {
...
+			p := pc.BasicAuth
+
+			caDest, caData := newSourceDataIDPConfigMap(i, p.CA.Name, corev1.ServiceAccountRootCAKey)
+			configMaps[caDest] = caData
+
+			clientCertDest, clientCertData := newSourceDataIDPSecret(i, p.TLSClientCert.Name, corev1.TLSCertKey)
+			secrets[clientCertDest] = clientCertData
+
+			clientKeyDest, clienKeyData := newSourceDataIDPSecret(i, p.TLSClientKey.Name, corev1.TLSPrivateKeyKey)
+			secrets[clientKeyDest] = clienKeyData
```
Comment (ericavonb) [inline pkg/operator2/configsync.go]: typo - `s/clienKeyData/clientKeyData/` (missing the 't' in 'client')

**Comment by stlaz** [comment]
typo fixed

**Comment by stlaz** [comment]
latest push fixes writing to `nil` maps

**Comment by stlaz** [comment]
I chipped away the refactoring so that we can merge this earlier. See https://github.com/openshift/cluster-osin-operator/pull/26 for the follow-up.

**Comment by stlaz** [comment]
Fixed bug in `newSourceDataIDPConfigMap()`, it had `secretVolume()` instead of the proper volume func.

**Comment by enj** [comment]
Tested with 4.0 cluster, htpasswd still works so this hopefully fine.

/retest

**Comment by enj** [comment]
`kube:admin` still works.  Time for the :hammer:

**Pull request merged**